### ISSUE TITLE
Project Masterplan

### ISSUE THREAD

**Issue by JMitnik** [lucasfijen/How-far-should-we-step-Policy-Improvement-with-Trust-Region-Optimisation#1]
# Description

```
Compare trust region policy optimization to the natural policy gradient.
TRPO allows the learning rate to adapt during learning. Do you observe
step size changes as learning progresses? (How) does this depend on the
environment and/or the policy parametrization?
```
# Things to do: Coding
1. Basic code setup
- [ ] Setup python basic boilerplate (use our previous code-base for this). Lucas/Guido's code

2. Experimental setup
- [ ] Configure config object (implement using `dataclass`)
    - This contains info about the environment, specific parameters for an experiment, your network
- [ ] Configure a basic neural network for our learned (use the same one as Lucas's code for instance)
- [ ] Configure training loop with basic optimizer
- [ ] Which metric interests us is step-size.

3. Generic Policy Gradient API
- [ ] Decide a generic interface for both Natural Policy Gradient and TRPO. It would be good if we can _swap_ easily between these methods.
    - Which methods do we need?
    - When do we call this exactly? Is this an `optimizer` or is this a `loss`?

4. Natural Policy Gradient API
- [ ] Decide more specific API for Natural Policy gradient. Which methods (specific to NPG) might we need on top?
- [ ] Write class which implements NPG
- [ ] Write some test to ensure that it works as expected.

5. TRPO API
- [ ] Decide more specific API for TRPO. Which methods (specific to TRPO) might we need on top?
- [ ] Write class which implements TRPO
- [ ] Write some test to ensure that it works as expected.

# Draft
- [x]  Motivation and research question (**Jonathan**)

**Algorithms**
- [ ]  Put in description of NPG and TRPO algorithms
- [ ]  Descripe our policy neural networks:
    - [ ]  Describe the current simple one
    - [ ]  Algorithm 2: think of one and describe it (use Deep RL that matters)

**Experimental design**: explain our approach
- [ ] Explain how we plan to vary the environments
- [ ] (Use RL Lecture 10 theory)
- [ ] Explain what we pay attention to: which metrics
- [ ] Can we tackle the random-seed problem? Namely, if we do X runs across X different random seeds, the output distributions are different enough that any result can essentially be generated (see "section Random seeds and Trail" in "Deep RL that Matters")

**Experimental design**: explaining hyperparameter selection
- [x]  Explain which hyperparameters we vary
- [x]  Explain how we choose hyperparameters
- [x] Explain whether we do reward scaling

**Experimental design**: explaining the environments:
- [x]  Explain the current one: why we chose this one
- [ ]  Select/think of  2 similar ones, and explain those
- [x]  Ensure we refer to all the environments using URL's or resources or something

# Things to do: Writing
- [ ] Decide on "Blog" name
- [ ] Decide on "Blog" platform (medium/Notion/own website?)
